---
title: "Case study of constructing cell type-specifc lncRNA regulatory networks in ASD across 17 cell types" 
author: "Chenchen Xiong"
date: '`r Sys.Date()`'
output:
    BiocStyle::html_document:
      toc: yes
    BiocStyle::pdf_document:
      toc: yes
vignette: >
    %\VignetteIndexEntry{Case study of constructing cell type-specifc lncRNA regulatory networks in ASD} 
    %\VignettePackage{Cycle} 
    % \VignetteEngine{knitr::rmarkdown} 
    % \usepackage[utf8]{inputenc} 
    % \VignetteEncoding{UTF-8}
---

```{r style, echo=FALSE, results="asis", message=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(tidy = FALSE,
    warning = FALSE,
    message = FALSE)
```


```{r echo=FALSE, results='hide', message=FALSE}
library(Seurat)
library(data.table)
library(readxl)

library(Rmagic)
library(viridis)
library(phateR)
library(harmony)
library(pracma)

library(foreach)
library(doParallel)
library(parallel)

library(ggplot2)
library(corrplot)
library(clusterProfiler)
library(enrichplot)

library(igraph)
library(miRspongeR)
library(mldr)
library(utiml)
library(e1071)
library(Cycle)
```


# Data preparation and preprocessing
```{r, eval=FALSE, include=TRUE}
# Raw count snRNA data could download from https://autism.cells.ucsc.edu/
# Load processed matched lncRNA and mRNA snRNA-seq data with ASD samples could be downloaded from https://drive.google.com/drive/folders/1_xjA_S77POIh28w49trPnuFUCTWagBzG

# Data preparation
## load raw single-cell count data
data_dir <- './rawMatrix/'
count_matrix <- Read10X(data.dir = data_dir)
ASD <- CreateSeuratObject(counts = count_matrix, project = "ASD", min.cells = 0, min.features = 0)

## load meta data
ASD_meta <- fread('./rawMatrix/meta.txt',sep = '\t',header = T,stringsAsFactors = F,check.names = F,data.table = F,encoding = 'UTF-8') # all 31 sample provided by paper
# Sample and clinical information for ASD and epilepsy individuals.
raw_meta <- read_excel('NIHMS1053005-supplement-Data_S1.xlsx',sheet = 'ASD_and_Control_samples')

## HGNC data
# https://www.genenames.org/download/statistics-and-files/
hgnc_complete_set <- read.delim("hgnc_complete_set.txt", stringsAsFactors=FALSE)

# 1.select ASD-related single-cell data and meta data
## select individual with ASD and without epilepsy, 7 unique individual in total
raw_meta_ASD <- subset(raw_meta, epilepsy=='no' & Diagnosis == 'ASD')
raw_meta_ASD2 <- merge(ASD_meta, raw_meta_ASD, by.x = 'individual', by.y = 'Individual') # get barcodes infor
## select subdata from seurat object and to dataframe, rows are cells and colnums are genes
ASD_sub_cells <- ASD_meta[which(ASD_meta$individual %in% unique(raw_meta_ASD2$individual)),1]
ASD_sub <- subset(ASD, cells = ASD_sub_cells)
ASD_sub_data <- as.data.frame(t(as.data.frame(ASD_sub@assays[["RNA"]]@counts))) # 24296 65217

# 2.Filtering data
## 2.1 keep genes expressed in at least 10 cells
keep_cols <- colSums(ASD_sub_data > 0) > 10
ASD_sub_data_filter <- ASD_sub_data[,keep_cols]

## 2.2 keep cells with at least 1000 UMIs
### refer to http://htmlpreview.github.io/?https://github.com/KrishnaswamyLab/MAGIC/blob/master/Rmagic/inst/examples/bonemarrow_tutorial.html, just keep cells with at least 1000 UMIs when some cells contains more than 15000 UMIs.
keep_rows <- rowSums(ASD_sub_data_filter) > 1000
ASD_sub_data_filter <- ASD_sub_data_filter[keep_rows,]

# 2.Normalizing data
## MAGIC and harmony all need normalized data, We use the normalizing method recommended by MAGIC. 
ASD_sub_data_normalize <- library.size.normalize(ASD_sub_data_filter)

## meta information with filter and normalizd data, and input for harmony
meta_data <- as.data.frame(raw_meta_ASD2[,1:5])
meta_data <- meta_data[match(rownames(ASD_sub_data_normalize),meta_data[,2]),] # 18883     5
meta_data <- as.data.frame(meta_data)

# 3. Running harmony to data integration, time consumption more than 2 hours
ASD_sub_data_harmony <- HarmonyMatrix(ASD_sub_data_normalize, meta_data,vars_use = c('individual'),do_pca = F)

# 4. Running MAGIC to data imputation
set.seed(0)
ASD_sub_data_MAGIC <- magic(ASD_sub_data_harmony, genes="all_genes",t='auto',knn=15)
ASD_sub_data_MAGIC <- as.data.frame(ASD_sub_data_MAGIC$result)

# 5. gene annotation using HGNC
gene.names <- colnames(ASD_sub_data_MAGIC$result)
index <- match(gene.names,hgnc_complete_set$symbol)
symbol <- hgnc_complete_set[index,c("symbol","locus_group","locus_type","entrez_id")]
geneName <- cbind(gene.names,symbol)
geneName <- na.omit(geneName)

hgnc_ncRNAs_all <- geneName[grep('long non-coding',geneName$locus_type),] # 1611    5
hgnc_mRNAs_all <- geneName[grep('gene with protein product',geneName$locus_type),] # 15608     5

# 6. Separate ncRNA and mRNA data after MAGIC
## 6.1 ncRNA data
ASD_MAGIC_ncR <-  intersect(colnames(ASD_sub_data_MAGIC),hgnc_ncRNAs_all[,2])
ASD_MAGIC_ncR_data <- ASD_sub_data_MAGIC[,ASD_MAGIC_ncR] 

## 6.2 mRNA data
ASD_MAGIC_mR <- intersect(colnames(ASD_sub_data_MAGIC),hgnc_mRNAs_all[,2])
ASD_MAGIC_mR_data <- ASD_sub_data_MAGIC[,ASD_MAGIC_mR] 

## 6.3 combine ncRNA data with mRNA data 
ASD_MAGIC_ncRmR_data <- cbind(ASD_MAGIC_ncR_data, ASD_MAGIC_mR_data) 

# 7. Statistics the number of cells for each cell type
ASD_MAGIC_ncRmR_data_meta <- ASD_meta[match(rownames(ASD_MAGIC_ncRmR_data), ASD_meta[,1]),]
ASD_MAGIC_ncRmR_data_cluster <- as.data.frame(table(ASD_MAGIC_ncRmR_data_meta$cluster))
ASD_MAGIC_ncRmR_data_cluster
#                Var1 Freq
# 1            AST-FB  338
# 2            AST-PP 2291
# 3       Endothelial  468
# 4             IN-PV 1007
# 5            IN-SST  999
# 6           IN-SV2C  447
# 7            IN-VIP 1362
# 8              L2/3 2496
# 9                L4 1278
# 10             L5/6  837
# 11          L5/6-CC  896
# 12        Microglia  351
# 13          Neu-mat  524
# 14       Neu-NRGN-I  956
# 15      Neu-NRGN-II  332
# 16 Oligodendrocytes 2182
# 17              OPC 2119

```

# Inferring cell type-specific lncRNA-mRNA regulatory networks in ASD across 17 cell types
```{r, eval=FALSE, include=TRUE}
load('ASD_MAGIC_ncRmR_data_cluster.rda')
load('ASD_MAGIC_ncR_mR_data.rda')
load('ASD_MAGIC_ncRmR_data_meta.rda')

celltypes = ASD_MAGIC_ncRmR_data_cluster[,1]

Cycle_networks = list()
for(celltype in celltypes){
  lncR_data <- ASD_MAGIC_ncR_data[ASD_MAGIC_ncRmR_data_meta[ASD_MAGIC_ncRmR_data_meta$cluster==celltype,1],]
  mR_data <-ASD_MAGIC_mR_data[ASD_MAGIC_ncRmR_data_meta[ASD_MAGIC_ncRmR_data_meta$cluster==celltype,1],]
  cell_num <- nrow(lncR_data)
  res_list <-  Cycle_network(lncR_data, mR_data, boxsize = 0.1, p.value.cutoff = 0.05, num.cores = 12, dev=TRUE, iteration=TRUE, cell_id=NULL, maxiter=20)
  overlap <- Overlap.lncRNet(net = res_list,overlap.num = cell_num,type = 'equal')
  overlap <- overlap$net.table
  overlap_90pct <- overlap[which(overlap[,2]>=round(cell_num*0.9)),]
  overlap_90pct_network <- overlap_90pct %>% separate(col = net.transform,into = c('lncRNAs','mRNAs'),sep = ' & ')
  Cycle_networks[[celltype]] <- overlap_90pct_network[,1:2]
}

```


# Identifying cell type-specific hub lncRNAs using Possion model
```{r, eval=FALSE, include=TRUE}

Cycle_hubs_lncRNAs <-  lapply(seq(Cycle_networks),function(i){
  Cycle_network_graph <-  make_graph(c(t(Cycle_networks[[i]][,1:2])),directed=TRUE)
	Cycle_network_degree <- degree(Cycle_network_graph,mode='out')
	Cycle_network_degree <- Cycle_network_degree[Cycle_network_degree>0]
  
  n <- length(Cycle_network_degree) 
  m <- nrow(Cycle_networks[[i]])
  p1 <- m/(n^2)
  lamda <- n*p1
  p <- ppois(Cycle_network_degree - 1, lambda = lamda, lower=FALSE)
  hub <- p[which(p < 0.05)]
  hub_names <- names(hub)
})

names(Cycle_hubs_lncRNAs) = names(Cycle_networks)

```


# Network topological analysis of cell type-specific lncRNA regulatory network
```{r, eval=FALSE, include=TRUE}

library(tidyr)
library(igraph)

edge_density_pvalue=c()
for (i in 1:length(Cycle_networks)) {

  g1 <- Cycle_networks[[i]]
  g1 <- make_graph(c(t(g1[, 1:2])),directed = T)
  
  m <- gsize(g1)
  n <- vcount(g1)
  l <- mean_distance(g1)
  d <- edge_density(g1)
  
  degree <- degree(g1,mode="all")
  fit <- fit_power_law(degree)
  
  pvalue_path_density <- Random_net_parallel(l, d, n, m, perm = 100, directed = TRUE, num.cores = 4)]
  pvalue_path_raw <- pvalue_path_density$pvalue_path_raw
  pvalue_path <- pvalue_path_density$pvalue_path
  pvalue_density_raw <- pvalue_path_density$pvalue_density_raw
  pvalue_density <- pvalue_path_density$pvalue_density
  
  network_attr_tm <- c(names(Cycle_networks)[i],m,n,l,d,pvalue_path_raw,pvalue_path,pvalue_density_raw,pvalue_density,fit$continuous,fit$alpha,fit$xmin,fit$logLik,fit$KS.stat,fit$KS.p)
  
  edge_density_pvalue <- rbind(edge_density_pvalue,network_attr_tmp)
}
colnames(edge_density_pvalue) <- c('cell types','#edges','#nodes','path','density','path_pvalue','path_pvalue(-log10)','density_pvalue','density_pvalue(-log(10))','fit$continuous','fit$alpha',"fit$xmin","fit$logLik","fit$KS.stat","fit$KS.p")
rownames(edge_density_pvalue) <- NULL


```


# Identifying cell-specific hub lncRNAs 
```{r, eval=FALSE, include=TRUE}
Cycle_hubs_lncRNAs = lapply(seq(Cycle_networks),function(i){
  
  Cycle_network_graph = make_graph(c(t(Cycle_networks[[i]][,1:2])),directed=TRUE)
	Cycle_network_degree <- degree(Cycle_network_graph,mode='out')
	Cycle_network_degree=Cycle_network_degree[Cycle_network_degree>0]
  
  n <- length(Cycle_network_degree) 
  m <- nrow(Cycle_networks[[i]])
  p1 <- m/(n^2)
  lamda <- n*p1
  p <- ppois(Cycle_network_degree - 1, lambda = lamda, lower=FALSE)
  hub <- p[which(p < 0.05)]
  hub_names <- names(hub)
})

```


# Calculating similarity matrix between two list of lncRNA-mRNA regulatory networks and hub lncRNAs
```{r, eval=FALSE, include=TRUE}

## Calculating similarity matrix of cell type-specific lncRNA-mRNA regulatory networks across cell types
Cycle_network_Sim <- Sim.network(Cycle_networks, Cycle_networks, directed = TRUE)
## Similarity plot in terms of cell type-specific lncRNA-mRNA regulatory netowork    
rownames(Cycle_network_Sim) <- colnames(Cycle_network_Sim) <- unlist(lapply(names(Cycle_networks), function(x){strsplit(x,split = '_')}[[1]][2]))

## uniqueness of lncRNA-mRNA regulatory networks across cell types
Cycle_network_uniueness <- 1-Cycle_network_Sim
corrplot(Cycle_network_uniueness, method = "pie", type = "lower", diag = FALSE)

## Calculating similarity matrix of cell type-specific hub lncRNAs across cell types
Cycle_hub_Sim <- Sim.hub(Cycle_hubs_lncRNAs, Cycle_hubs_lncRNAs)
## uniqueness of lncRNA-mRNA regulatory networks across cell types
rownames(Cycle_hub_Sim) <- colnames(Cycle_hub_Sim) <- unlist(lapply(names(Cycle_networks), function(x){strsplit(x,split = '_')}[[1]][2]))
Cycle_hub_Sim_uniueness <- 1-Cycle_hub_Sim
corrplot(Cycle_hub_Sim_uniueness, method = "pie", type = "lower", diag = FALSE)

```


# Discovering conserved and rewired lncRNA-mRNA regulatory networks between 17 cell types
```{r, eval=FALSE, include=TRUE}

## Overlap of multiple hubs
Overlap_net.all <- Overlap.net(Cycle_networks, overlap.num = 1, type = "least") 
Overlap_net.1 <- Overlap.net(Cycle_networks, overlap.num = 1, type = "equal")
Overlap_net.2 <- Overlap.net(Cycle_networks, overlap.num = 2, type = "least")
Overlap_net.5 <- Overlap.net(Cycle_networks, overlap.num = 5, type = "least")
Overlap_net.9 <- Overlap.net(Cycle_networks, overlap.num = 9, type = "least")
Overlap_net.17 <- Overlap.net(Cycle_networks, overlap.num = 17, type = "equal")

## Overlap of multiple hubs
Overlap.hub.all <- Overlap.hub(Cycle_hubs_lncRNAs, overlap.num = 1, type = "least") 
Overlap.hub.1 <- Overlap.hub(Cycle_hubs_lncRNAs, overlap.num = 1, type = "equal")
Overlap.hub.2 <- Overlap.hub(Cycle_hubs_lncRNAs, overlap.num = 2, type = "least")
Overlap.hub.5 <- Overlap.hub(Cycle_hubs_lncRNAs, overlap.num = 5, type = "least")
Overlap.hub.9 <- Overlap.hub(Cycle_hubs_lncRNAs, overlap.num = 9, type = "least")
Overlap.hub.17 <- Overlap.hub(Cycle_hubs_lncRNAs, overlap.num = 17, type = "equal")

Overlap_Cycle_network_rewired <- Overlap_net.1
Overlap_Cycle_hubs_rewired <- Overlap.hub.1
Overlap_Cycle_hubs_conserved <- Overlap.hub.17

network_overlap_num <- c(nrow(Overlap_net.1)/nrow(Overlap_net.all),nrow(Overlap_net.2)/nrow(Overlap_net.all),nrow(Overlap_net.5)/nrow(Overlap_net.all),nrow(Overlap_net.9)/nrow(Overlap_net.all),nrow(Overlap_net.17)/nrow(Overlap_net.all))
hub_overlap_num <- c(length(Overlap.hub.1)/length(Overlap.hub.all),length(Overlap.hub.2)/length(Overlap.hub.all),length(Overlap.hub.5)/length(Overlap.hub.all),length(Overlap.hub.9)/length(Overlap.hub.all),length(Overlap.hub.17)/length(Overlap.hub.all))

radar_plot_data <- rbind(hub_overlap_num,network_overlap_num)
colnames(radar_plot_data) <- c('=1','>=2','>=5','>=9','=17')
rownames(radar_plot_data) <- c('Hub lncRNAs','Interactions')

max_min <- data.frame('=1'=c(1,0),'>=2'=c(1,0),'>=5'=c(1,0),'>=9'=c(1,0),'=17'=c(1,0))
rownames(max_min) <- c("Max", "Min")
colnames(max_min) <- colnames(radar_plot_data)

radar_plot_data <- rbind(max_min, radar_plot_data)

radar  <-  {
  fmsb::radarchart(
  radar_plot_data, axistype = 1,
  # Customize the polygon
  pcol = c ("purple", "cyan"), pfcol = scales::alpha(c("purple", "cyan"),0.5), plwd = 3, plty = 1,
  # Customize the grid
  cglcol = "black", cglty = 1, cglwd =3,
  # Customize the axis
  axislabcol = "red",
  # Variable labels
  vlcex = 2, vlabels = colnames(radar_plot_data),
  caxislabels = c('0%', "25%", "50%", "75%", "100%"))
# Add an horizontal legend
legend(
  x = "bottom", legend = rownames(radar_plot_data[-c(1,2),]), horiz = TRUE,
  bty = "n", pch = 22 , col = c("purple", "cyan"),
  text.col = "black", cex = 1, pt.cex = 1.5
  )
  recordPlot()
}
print(radar)
ggsave(plot = replayPlot(radar),filename = 'radar.tiff',dpi = 600,height = 5,width = 5,bg='transparent')
```


# Performance of rewired lncRNA-mRNA regulatory networks and hub lncRNAs for classifying cell types
```{r, eval=FALSE, include=TRUE}

cell_subtype <- ASD_MAGIC_ncRmR_data_meta[,1:2]
cell_subtype[,2] <- gsub('[/]','',cell_subtype[,2])
cell_subtype[,2] <- gsub('[-]','',cell_subtype[,2])
Overlap_Cycle_network_rewired_genes <- list(unique(c(Overlap_Cycle_network_rewired[,1],Overlap_Cycle_network_rewired[,2])))
Overlap_Cycle_hubs_rewired_genes <- list(Overlap_Cycle_hubs_rewired)

## Performance of rewired lncRNA-mRNA regulatory networks
## br strategt
rewired_network_classify_br <- classify(lncRExp = ASD_MAGIC_ncR_data,mRExp = ASD_MAGIC_mR_data,celltype = cell_subtype,genelist = Overlap_Cycle_network_rewired_genes, method = "br", base.algorith = "SVM", cv.folds = 10, cv.sampling = "stratified", cv.seed = 12345)
## baseline strategy
rewired_network_classify_baseline <- classify(lncRExp = ASD_MAGIC_ncR_data,mRExp = ASD_MAGIC_mR_data,celltype = cell_subtype,genelist = Overlap_Cycle_network_rewired_genes, method = "baseline", base.algorith = "SVM", cv.folds = 10, cv.sampling = "stratified", cv.seed = 12345)
## RANDOM method
rewired_network_classify_random <- classify(lncRExp = ASD_MAGIC_ncR_data,mRExp = ASD_MAGIC_mR_data,celltype = cell_subtype,genelist = Overlap_Cycle_network_rewired_genes, method = "br", base.algorith = "RANDOM", cv.folds = 10, cv.sampling = "stratified", cv.seed = 12345)

## Performance of rewired hub lncRNAs
## br strategy
rewired_hubs_classify_br <- classify(lncRExp = ASD_MAGIC_ncR_data,mRExp = ASD_MAGIC_mR_data,celltype = cell_subtype,genelist = Overlap_Cycle_hubs_rewired_genes, method = "br", base.algorith = "SVM", cv.folds = 10, cv.sampling = "stratified", cv.seed = 12345)
## baseline strategt
rewired_hubs_classify_baseline <- classify(lncRExp = ASD_MAGIC_ncR_data,mRExp = ASD_MAGIC_mR_data,celltype = cell_subtype,genelist = Overlap_Cycle_hubs_rewired_genes, method = "baseline", base.algorith = "SVM", cv.folds = 10, cv.sampling = "stratified", cv.seed = 12345)
## RANDOM method
rewired_hubs_classify_random <- classify(lncRExp = ASD_MAGIC_ncR_data,mRExp = ASD_MAGIC_mR_data,celltype = cell_subtype,genelist = Overlap_Cycle_hubs_rewired_genes, method = "br", base.algorith = "RANDOM", cv.folds = 10, cv.sampling = "stratified", cv.seed = 12345)

```


# Enrichment analysis Of rewired lncRNA-mRNA regulatory network
```{r, eval=FALSE, include=TRUE}

rewired_network_list <- list(unique(c(unique(Overlap_Cycle_network_rewired[,1]),unique(Overlap_Cycle_network_rewired[,2]))))
rewired_network_list_FEA <- moduleFEA(rewired_network_list)
rewired_network_FEA_GO <- dotplot(rewired_network_list_FEA[[1]][[1]],showCategory = 20,label_format=50)
rewired_network_FEA_KEGG <- dotplot(rewired_network_list_FEA[[2]][[1]],showCategory = 20,label_format=100)
rewired_network_FEA_Reactome <- dotplot(rewired_network_list_FEA[[3]][[1]],showCategory = 20,label_format=100)

cell_markers  <- read.xlsx("D:/Z_project/2_LCSlncR_modified/0_data/5.biomarkers_data/Cell_marker_Human_seq_cellMarker2.xlsx") %>% tidyr::unite("marker", tissue_type, cancer_type, cell_name, sep=", ") %>% dplyr::select(marker, Symbol)   
Overlap_LCSlncR_networks_rewired_gene = unique(c(unique(Overlap_LCSlncR_networks_rewired[,1]),Overlap_LCSlncR_networks_rewired[,2]))
rewired_network_cellmarker_FEA = enricher(Overlap_LCSlncR_networks_rewired_gene,TERM2GENE=cell_markers)
```


# cell communication between 17 cell types
```{r, eval=FALSE, include=TRUE}

## Identifying crosstalk network in terms of network similarity matrix between 17 cell types
    Cycle_network_adjacency_matrix <- ifelse(Cycle_network_Sim > median(Cycle_network_Sim[lower.tri(Cycle_network_Sim)]), 1, 0)
    diag(Cycle_network_adjacency_matrix) <- 0
    colnames(Cycle_network_adjacency_matrix) <- rownames(Cycle_network_adjacency_matrix) <- unlist(lapply(names(Cycle_networks),function(x){strsplit(x,split = '_')[[1]][2]}))
    Cycle_network_adjacency_matrix_graph <- graph_from_adjacency_matrix(Cycle_network_adjacency_matrix, mode = "undirected")
    
## Identifying crosstalk network in terms of hub lncRNA similarity matrix between 17 cell types   
    Cycle_hub_adjacency_matrix <- ifelse(Cycle_hub_Sim > median(Cycle_hub_Sim[lower.tri(Cycle_hub_Sim)]), 1, 0)
    diag(Cycle_hub_adjacency_matrix) <- 0
    colnames(Cycle_hub_adjacency_matrix) <- rownames(Cycle_hub_adjacency_matrix) <-  unlist(lapply(names(Cycle_networks),function(x){strsplit(x,split = '_')[[1]][2]}))
    Cycle_hub_adjacency_matrix_graph <- graph_from_adjacency_matrix(Cycle_hub_adjacency_matrix, mode = "undirected")

```


# Session information
```{r}
sessionInfo()
```

